src/agent_workload.py

```diff
@@ -0,0 +1,185 @@
+                    "override": "replace",
+                    "summary": "agent",
+                    "command": f"/bin/agent {self._cli_args()}",
+                    "startup": "enabled",
+                },
+            },
+        }
+        self._container.add_layer(self._service_name, pebble_layer, combine=True)
+        self._container.autostart()
+
+        if version := self.version:
+            self._unit.set_workload_version(version)
+        else:
+            logger.debug(
+                "Cannot set workload version at this time: could not get grafana-agent version."
+            )
+
+    def is_ready(self):
+        """Is ready."""
+        return self._container.can_connect()
+
+    @property
+    def version(self) -> Optional[str]:
+        """Returns the version of the agent.
+
+        Returns:
+            A string equal to the agent version
+        """
+        if not self.is_ready:
+            return None
+
+        # Output looks like this:
+        # agent, version v0.26.1 (branch: HEAD, revision: 2b88be37)
+        version_output, _ = self._container.exec(["/bin/agent", "-version"]).wait_output()
+        result = re.search(r"v(\d*\.\d*\.\d*)", version_output)
+        return result.group(1) if result else None
+
+    def read_file(self, filepath: Union[str, pathlib.Path]):
+        """Read a file's contents.
+
+        Returns:
+            A string with the file's contents
+        """
+        return self._container.pull(filepath).read()
+
+    def write_file(self, path: Union[str, pathlib.Path], text: str) -> None:
+        """Write text to a file.
+
+        Args:
+            path: file path to write to
+            text: text to write to the file
+        """
+        self._container.push(path, text, make_dirs=True)
+
+    def restart(self) -> None:
+        """Restart grafana agent."""
+        self._container.restart(self._service_name)
+
+    def _update_config(self) -> None:
+        if not self.is_ready:
+            # Workload is not yet available so no need to update config
+            self.status.config = WaitingStatus("Workload is not yet available")
+            return
+
+        config = self._render_config()  # TODO: Must not be None
+        assert config is not None
+
+        try:
+            old_config = self.read_file(self.CONFIG_PATH)
+            logger.info("Old config: %s", old_config)
+            old_config = yaml.safe_load(old_config)
+        except (FileNotFoundError, PathError, ParserError, ConstructorError):
+            # The file does not yet exist?
+            old_config = None
+
+        if config == old_config:
+            # Nothing changed, possibly new installation. Move on.
+            self.status.config = ActiveStatus()
+            return
+
+        try:
+            self.write_file(self.CONFIG_PATH, yaml.dump(config))
+        except APIError as e:
+            logger.warning(str(e))
+            self.status.config = WaitingStatus(str(e))
+            return
```

Comment:
issue: I do not think waiting status makes sense. We are not waiting on a related app. In fact, this error is not resolvable at all, so we should surface the exception in order to go in to error state.

Reply:
This is [copied](https://github.com/canonical/grafana-agent-k8s-operator/blob/5be2d358b4a2fbb4dca41625626b0a9624ca11d1/src/grafana_agent.py#L418) from gagent.

Seems like:
- It should be WaitingStatus if we're before pebble-ready ("auto resolvable")
- BlockedStatus otherwise

So I agree, all in all this should be BlockedStatus.